### PR TITLE
Compactor: Add metrics for retention progress

### LINKED ISSUE #4835: Compact: Retention progress metrics
### Is your proposal related to a problem?

Because retention happens after compaction and downsampling, so it might get huge retention backlog to do in the object storage. We also got the same problem mentioned in https://github.com/thanos-io/thanos/pull/4406.

It would be great to have a way to tell how many blocks are beyond their retention period but don't get deleted.

### Describe the solution you'd like

We have introduced the progress calculator to get compaction and downsampling progress as metrics in https://github.com/thanos-io/thanos/pull/4801.

We can implement another `RetentionProgressCalculator` to do this work.

### PR DESCRIPTION
**Pull request by metonymic-smokey**
Fixes #4835 

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
Added a retention progress calculator, similar to the ones for compaction and downsampling.

### CODE REVIEW

**Comment by yeya24** [comment]
Change is relevant to users. Can you update the changelog?

File: pkg/compact/compact.go [hunk 1]
```diff
@@ -667,6 +667,55 @@ func (ds *DownsampleProgressCalculator) ProgressCalculate(ctx context.Context, g
 	return nil
 }
 
+// RetentionProgressMetrics contains Prometheus metrics related to retention progress.
+type RetentionProgressMetrics struct {
+	NumberOfBlocksDeleted *prometheus.GaugeVec
```
Comment (yeya24) [inline pkg/compact/compact.go:672]: `NumberOfBlocksDeleted` uses past tense but the progress is future tense. Let's rename it to `NumberOfBlocksToDelete` or `NumberOfBlocksRetention`?

File: pkg/compact/compact.go [hunk 2]
```diff
@@ -667,6 +667,55 @@ func (ds *DownsampleProgressCalculator) ProgressCalculate(ctx context.Context, g
...
+	retentionByResolution map[ResolutionLevel]time.Duration
+}
+
+// NewRetentionProgressCalculator creates a new RetentionProgressCalculator.
+func NewRetentionProgressCalculator(reg prometheus.Registerer, retentionByResolution map[ResolutionLevel]time.Duration) *RetentionProgressCalculator {
+	return &RetentionProgressCalculator{
+		retentionByResolution: retentionByResolution,
+		RetentionProgressMetrics: &RetentionProgressMetrics{
+			NumberOfBlocksDeleted: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_compact_todo_deleted_blocks",
```
Comment (yeya24) [inline pkg/compact/compact.go:687]: Same here. What about `thanos_compact_todo_retention_blocks`?

**Review by yeya24 — DISMISSED** [review]
LGTM. Only some small nits about naming. Let's mark this pr for review.

File: pkg/compact/compact.go (hunk 2 above)
Comment (metonymic-smokey) [reply pkg/compact/compact.go:687]: @yeya24 I think `deletion_blocks` is more suitable than `retention_blocks`.

File: pkg/compact/compact_test.go [hunk 3]
```diff
@@ -205,6 +205,148 @@ func createBlockMeta(id uint64, minTime, maxTime int64, labels map[string]string
 	return m
 }
 
+func TestRetentionProgressCalculate(t *testing.T) {
+	logger := log.NewNopLogger()
+	reg := prometheus.NewRegistry()
+	unRegisterer := &receive.UnRegisterer{Registerer: reg}
```
Comment (yeya24) [inline pkg/compact/compact_test.go:211]: I believe we don't need to use unregisterer. Can you also remove it in the compact and downsample tests?

Comment (metonymic-smokey) [reply pkg/compact/compact_test.go]: Thanks!
Could you PTAL at my latest commit and let me know if that's what's expected?

**Review by yeya24 — APPROVED** [review]
LGTM

**Pull request merged**